### PR TITLE
CDAP-6794 Support cross namespace write for map reduce program through Output class

### PR DESCRIPTION
**Pull request by lumig242**
Support cross namespace write for map reduce program through `Output` class. 

Use the API as:  `context.addOutput(Output.ofDataset(datasetName).fromNamespace(namespace))`.

Similar to do `Input` change.

JIRA: https://issues.cask.co/browse/CDAP-6794
Build: http://builds.cask.co/browse/CDAP-DUT4550

### CODE REVIEW

File: cdap-api/src/main/java/co/cask/cdap/api/data/batch/Output.java [hunk 1]
```diff
@@ -104,6 +105,17 @@ private DatasetOutput(String name, Map<String, String> arguments) {
...
     }
+
+    public DatasetOutput fromNamespace(String namespace) {
+      DatasetOutput datasetOutput = new DatasetOutput(super.name, arguments);
+      datasetOutput.namesapce = namespace;
+      return datasetOutput;
+    }
+
+    @Nullable
+    public String getNamesapce() {
```
Comment (bdmogal) [inline cdap-api/src/main/java/co/cask/cdap/api/data/batch/Output.java]: typo

File: cdap-api/src/main/java/co/cask/cdap/api/data/batch/Output.java [hunk 2]
```diff
@@ -66,17 +67,16 @@ public Output alias(String alias) {
    *
    * @param datasetName the name of the output dataset
    */
-  public static Output ofDataset(String datasetName) {
+  public static DatasetOutput ofDataset(String datasetName) {
```
Comment (bdmogal) [inline cdap-api/src/main/java/co/cask/cdap/api/data/batch/Output.java:70]: is this backward-compatible?

File: cdap-api/src/main/java/co/cask/cdap/api/data/batch/Output.java [hunk 3]
```diff
@@ -95,6 +95,7 @@ public static Output of(String outputName, OutputFormatProvider outputFormatProv
   public static class DatasetOutput extends Output {
 
     private final Map<String, String> arguments;
+    private String namesapce;
```
Comment (bdmogal) [inline cdap-api/src/main/java/co/cask/cdap/api/data/batch/Output.java]: typo

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceTaskContext.java [hunk 4]
```diff
@@ -209,21 +209,32 @@ private void initializeTransactionAwares() {
     }
   }
 
-  /**
-   * This delegates the instantiation of the dataset to the super class, but in addition, if
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceTaskContext.java:213]: why remove this?

Comment (lumig242) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceTaskContext.java:213]: The piece of code starting the transaction is duplicate. So I extract them as a new function and move the comments there: https://github.com/caskdata/cdap/pull/6239/files#diff-aad06fed605dfd07f3806e9d07a5d3c2R228

File: cdap-api/src/main/java/co/cask/cdap/api/data/batch/Output.java (hunk 2 above)
Comment (lumig242) [inline cdap-api/src/main/java/co/cask/cdap/api/data/batch/Output.java:70]: I think it is. We do the same change for the `Input` class.

Al the usage of these `ofDataset` comes from the `addOutput()` method, which takes a `Output` as argument. Without this change, user will have to do an additional type conversion from `Output` -> `DatasetOutput` when specifying the namespace.

File: cdap-api/src/main/java/co/cask/cdap/api/data/batch/Output.java (hunk 3 above)
Comment (lumig242) [inline cdap-api/src/main/java/co/cask/cdap/api/data/batch/Output.java]: OH. Why I keep mis-spelling `namespace`.

File: cdap-api/src/main/java/co/cask/cdap/api/data/batch/Output.java [hunk 5]
```diff
@@ -104,6 +105,17 @@ private DatasetOutput(String name, Map<String, String> arguments) {
     public Map<String, String> getArguments() {
       return arguments;
     }
+
+    public DatasetOutput fromNamespace(String namespace) {
+      DatasetOutput datasetOutput = new DatasetOutput(super.name, arguments);
```
Comment (anwar6953) [inline cdap-api/src/main/java/co/cask/cdap/api/data/batch/Output.java:110]: can replace `super.name` with `name`?

File: cdap-unit-test/src/test/java/co/cask/cdap/test/app/DatasetCrossNSAccessWithMPApp.java [hunk 6]
```diff
@@ -29,12 +29,13 @@
 /**
  * App which copies data from one KVTable to another using a MapReduce program.
  */
-public class DatasetFromOtherSpaceWithMPApp extends AbstractApplication {
+public class DatasetCrossNSAccessWithMPApp extends AbstractApplication {
```
Comment (anwar6953) [inline cdap-unit-test/src/test/java/co/cask/cdap/test/app/DatasetCrossNSAccessWithMPApp.java]: curious - what is MP?

File: cdap-api/src/main/java/co/cask/cdap/api/data/batch/Output.java (hunk 5 above)
Comment (lumig242) [inline cdap-api/src/main/java/co/cask/cdap/api/data/batch/Output.java:110]: name is non-static and cannot be accessed in this class.

File: cdap-unit-test/src/test/java/co/cask/cdap/test/app/DatasetCrossNSAccessWithMPApp.java (hunk 6 above)
Comment (lumig242) [inline cdap-unit-test/src/test/java/co/cask/cdap/test/app/DatasetCrossNSAccessWithMPApp.java]: Oh typo. Should be "MAP". Changed

**Comment by lumig242** [comment]
Tested on a cluster.

**Comment by bdmogal** [comment]
LGTM. Squash and merge, if build is successful

**Comment by lumig242** [comment]
Will merge. This pr works for CDAP-6794. And plus, the changes here will also fix the CDAP-6835 bug.

**Comment by rohitsinha54** [comment]
@lumig242 : 

> Will merge. This pr works for CDAP-6794. And plus, the changes here will also fix the CDAP-6835 bug.

Did  someone reviewed the code changes to fix CDAP-6835 ? Also, I don't see another commit for fixing CDAP-6835 ?

**Comment by lumig242** [comment]
@rohitsinha54 . Showed this fix to @bdmogal in person. 

The related code segment was here: https://github.com/caskdata/cdap/pull/6239/files#diff-aad06fed605dfd07f3806e9d07a5d3c2R220 Added a method to override the `getdataset` from abstract class and start a transaction for the dataset processed.

**Pull request merged**